src/ActiveRoomsProvider.tsx

```diff
@@ -0,0 +1,85 @@
+import * as React from 'react';
+import { identity, sortBy } from 'lodash-es';
+import { addGlobalHook } from './common';
+import { PusherRoom } from './interfaces';
+
+const { withChatkit } = require('@pusher/chatkit-client-react');
+
+interface CreateRoomParams {
+  id?: string,
+  name: string,
+}
+
+export interface ActiveRoomsState {
+  loading: boolean,
+  activeRooms: PusherRoom[],
+  joinableRooms: PusherRoom[],
+  createRoom(room: CreateRoomParams): void,
+}
+
+export const ActiveRoomsContext = React.createContext<ActiveRoomsState>({
+  loading: true,
+  activeRooms: [],
+  joinableRooms: [],
+  createRoom: identity,
+});
+
+const sortRooms = (rooms: PusherRoom[]) => sortBy(rooms, ({ unreadCount, name }) => `${9999 - unreadCount}-${name}`);
+
+interface Props {
+  chatkit: any,
+  children: React.ReactChild,
+}
+
+export const ActiveRoomsProvider = withChatkit(({ chatkit, children }: Props) => {
+  const { currentUser, isLoading } = chatkit;
+  if (isLoading || !currentUser) {
+    return null;
+  }
+
+  const [loading, setLoading] = React.useState<boolean>(true);
+  const [activeRooms, setActiveRooms] = React.useState<PusherRoom[]>(sortRooms(currentUser.rooms));
+  const [joinableRooms, setJoinableRooms] = React.useState<PusherRoom[]>([]);
+  React.useEffect(() => {
+    chatkit.currentUser.enablePushNotifications();
+
+    const getRooms = async () => {
+      const joinableRooms = await currentUser.getJoinableRooms();
+      setJoinableRooms(joinableRooms);
+      setLoading(false);
+    };
+    getRooms();
+
+    const onAddedToRoom = (addedRoom: PusherRoom) => {
+      const sortedRooms = sortRooms([addedRoom, ...activeRooms]);
+      setActiveRooms(sortedRooms);
+    };
+    addGlobalHook(chatkit, 'onAddedToRoom', onAddedToRoom);
+
+    const onRemovedFromRoom = (removedRoom: PusherRoom) => {
+      setActiveRooms(activeRooms.filter(room => room.id !== removedRoom.id));
+    };
+    addGlobalHook(chatkit, 'onRemovedFromRoom', onRemovedFromRoom);
+
+    const onRoomUpdated = (updatedRoom: PusherRoom) => {
+      const newRooms = activeRooms.filter(room => room.id !== updatedRoom.id);
+      newRooms.push(updatedRoom);
+      const sortedRooms = sortRooms(newRooms);
+      setActiveRooms(sortedRooms);
+    };
+    addGlobalHook(chatkit, 'onRoomUpdated', onRoomUpdated);
```

Comment:
I'd split these into multiple `useEffect` invocations with separate cleanup methods, it helps separate different functionality blocks.

Reply:
👍

Reply:
They're all so similar... I'm making a useGlobalHook hook for this.